Translate react without es6 (#105)
reactjs/de.reactjs.org: content/docs/react-without-es6.md

---
id: react-without-es6

> title: React Ohne ES6

Reviewer: ```suggestion
title: React ohne ES6
```

permalink: docs/react-without-es6.html
---

> Normalerweise definiert man eine React Komponente als eine einfache JavaScript-Klasse:

Reviewer: ```suggestion
Normalerweise definiert man eine React-Komponente als eine einfache JavaScript-Klasse:
```

```javascript
class Greeting extends React.Component {
  render() {
    return <h1>Hallo, {this.props.name}</h1>;
  }
}
```

Wenn du ES6 noch nicht verwendest, kannst du stattdessen das `create-react-class` Modul verwenden:

```javascript
var createReactClass = require('create-react-class');
var Greeting = createReactClass({
  render: function() {
    return <h1>Hallo, {this.props.name}</h1>;
  }
});
```

Die API von ES6-Klassen ähnelt sich, mit wenigen Ausnahmen, der von `createReactClass()`.

> ## Default Props Deklarieren {#declaring-default-props}

Reviewer: I'm currently not sure,
if, in this case, `Deklarieren` is an `objektiviertes Verb`.

`(Das) Deklarieren von Default Props` - yes.

Can you give a sources with examples?

@ph1p, what's your opinion on this?

Author: I thought more along the lines of titleizing that sentence, that's where the uppercase comes from. Happy to change it to whatever you guys think makes more sense

Reviewer 2: `Deklarieren von Default Props` sounds right to me.

Author: Sorry for the delay. I like that change, I'll go ahead and fix that now.

});
```

> ## Den Anfangs-**state** Festsetzen {#setting-the-initial-state}

Reviewer: Same issue as `Deklarieren` in Line 30.

Author: Same as above

Reviewer 2: I think `Setzen des initialen States` sounds better.

Author: Same as above

> In ES6-Klassen kannst du den Anfangs-**state** definieren, indem du im **constructor** `this.state` zuweist:

Reviewer 2: ```suggestion
In ES6-Klassen kannst du den initialen State definieren, indem du im Konstruktor `this.state` zuweist:
```
If we change backticks or add bold text, the automatic merge tool might get a little confused. I think this can happen later, when the translation is done and we have the second, third.... Iteration.

> In ES6-Klassen kannst du den Anfangs-**state** definieren, indem du in dem **constructor** `this.state` zuweist:

Reviewer: ```suggestion
In ES6-Klassen kannst du den Anfangs-**state** definieren, indem du im **constructor** `this.state` zuweist:
```

}
```

> Mit `createReactClass()` müsst du eine separate `getInitialState` Methode bereitstellen, die den Anfangs-**state** zurückgibt:

Reviewer: ```suggestion
Mit `createReactClass()` musst du eine separate `getInitialState`-Methode bereitstellen, die den Anfangs-**state** zurückgibt:
```

> Mit `createReactClass()` musst du eine separate `getInitialState`-Methode bereitstellen, die den Anfangs-**state** zurückgibt:

Reviewer 2: ```suggestion
Mit `createReactClass()` musst du eine separate `getInitialState`-Methode bereitstellen, die den initialen State zurückgibt:
```

## Autobinding {#autobinding}

> In React Komponenten, die als ES6-Klassen deklarierten wurden, folgen Methoden der gleichen Semantik wie reguläre ES6-Klassen. Dies bedeutet, dass `this` sich nicht automatisch an die Instanz bindet. Du musst im **constructor** explizit `.bind(this)` verwenden:

Reviewer: ```suggestion
In React-Komponenten, die als ES6-Klassen deklariert wurden, folgen Methoden der gleichen Semantik wie reguläre ES6-Klassen. Dies bedeutet, dass `this` sich nicht automatisch an die Instanz bindet. Du musst im **constructor** explizit `.bind(this)` verwenden:
```

```javascript

> class SagHallo extends React.Component {

Reviewer: ```suggestion
class SayHello extends React.Component {
```

We talked about how to go with changing names [here](https://github.com/reactjs/de.reactjs.org/issues/4#issuecomment-467412690). 

> Translate comments and plaintext, but not function names, components or class names.

One missed changed in a function/component/class/variable name could break stuff.

@ph1p, I think you saw most of the code changes, what's your opinion on this?

  constructor(props) {
    super(props);

>     this.state = {nachricht: 'Hallo!'};

Reviewer 2: ```suggestion
    this.state = {message: 'Hallo!'};
```

    // Diese Zeile ist wichtig!
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {

>     alert(this.state.nachricht);

Reviewer 2: ```suggestion
    alert(this.state.message);
```

  }

  render() {
    // Da `this.handleClick` gebunden ist, können wir es als Event-Handler verwenden.
    return (
      <button onClick={this.handleClick}>
        Sag Hallo
      </button>
    );
  }
}
```

Mit `createReactClass()` ist dies nicht notwendig, da es alle Methoden bindet:

```javascript

> var SagHallo = createReactClass({

Reviewer 2: ```suggestion
var SayHello = createReactClass({
```

  getInitialState: function() {

>     return {nachricht: 'Hallo!'};

Reviewer 2: ```suggestion
    return {message: 'Hallo!'};
```

  },

  handleClick: function() {

>     alert(this.state.nachricht);

Reviewer 2: ```suggestion
    alert(this.state.message);
```

  },

  render: function() {
    return (
      <button onClick={this.handleClick}>
        Sag Hallo
      </button>
    );
  }
});
```

> Das bedeutet, dass ES6-Klassen mit etwas mehr Code für Event-Handler geliefert werden, aber die Leistung bei großen Anwendungen ist etwas besser.

Reviewer: ```suggestion
Das bedeutet, dass ES6-Klassen mit etwas mehr Code für Event-Handler geliefert werden, aber die Leistung bei großen Anwendungen etwas besser ist.
```

Wenn der Boilerplate-Code für dich zu unattraktiv ist, kannst du den **experimentellen** Syntaxvorschlag [Class Properties](https://babeljs.io/docs/plugins/transform-class-properties/) mit Babel aktivieren:

```javascript

> class SagHallo extends React.Component {

Reviewer 2: ```suggestion
class SayHello extends React.Component {
```

  constructor(props) {
    super(props);

>     this.state = {nachricht: 'Hallo!'};

Reviewer 2: ```suggestion
    this.state = {message: 'Hallo!'};
```

  }
  // WARNUNG: Diese Syntax ist experimentell!

>   // Die Verwendung eines Pfeils hier bindet die Methode:

Reviewer 2: ```suggestion
  // Die Verwendung eines Pfeils bindet hier die Methode:
```

  handleClick = () => {

>     alert(this.state.nachricht);

Reviewer 2: ```suggestion
    alert(this.state.message);
```
We do not translate code

  }

  render() {
    return (
      <button onClick={this.handleClick}>
        Sag Hallo
      </button>
    );
  }
}
```

Bitte beachte, dass die obige Syntax **experimentell** ist und sich die Syntax möglicherweise ändert oder der Vorschlag nicht in die Sprache eingebaut wird.

> Wenn du lieber auf Nummer sicher gehen möchten, hast du einige Möglichkeiten:

Reviewer: ```suggestion
Wenn du lieber auf Nummer sicher gehen möchtest, hast du einige Möglichkeiten:
```

> * Binde Methoden im **constructor**.

Reviewer 2: ```suggestion
* Binde Methoden im Konstruktor.
```

* Verwende Pfeilfunktionen, z.B. `onClick={(e) => this.handleClick(e)}`.
* Verwende weiterhin `createReactClass`.

## Mixins {#mixins}

>**Hinweis:**
>
>ES6 hat keine Unterstützung für Mixins, daher sind Mixins nicht unterstützt wenn du React zusammen mit ES6-Klassen benutzt.
>

> >**Wir haben auch zahlreiche Probleme in Projekten festgestellt, die Mixins verwenden, [und empfehlen, diese nicht im neuen Code zu verwenden](/blog/2016/07/13/mixins-considered-harmful.html).**

Reviewer: ```suggestion
>**Wir haben auch zahlreiche Probleme in Projekten festgestellt, die Mixins verwenden, [und empfehlen, diese nicht in neuem Code zu verwenden](/blog/2016/07/13/mixins-considered-harmful.html).**
```

>
>Dieser Abschnitt dient nur als Referenz.

Manchmal haben sehr unterschiedliche Komponenten ähnliche Funktionen. Diese werden manchmal als [Cross-Cutting Concern](https://de.wikipedia.org/wiki/Cross-Cutting_Concern) bezeichnet. Mit `createReactClass` kannst du dafür ein altes `mixins`-System verwenden.

> Ein häufiger Anwendungsfall ist eine Komponente, die sich in einem bestimmten Zeitintervall selbst aktualisieren möchte. Es ist einfach, `setInterval()` zu verwenden, aber es ist wichtig, das Intervall abzubrechen, wenn du es nicht mehr benötigst, um Speicherplatz zu sparen. React bietet [Lebenszyklusmethoden](/docs/react-component.html#the-component-lifecycle) an, die dich wissen lassen, wann eine Komponente im Begriff ist erstellt oder zerstört zu werden. Wir können ein Mixin erstellen, dass diese Methoden verwendet, um eine einfache `setInterval()`-Funktion bereitzustellen, die automatisch bereinigt wird, wenn deine Komponente zerstört wird.

Reviewer: ```suggestion
Ein häufiger Anwendungsfall ist eine Komponente, die sich in einem bestimmten Zeitintervall selbst aktualisieren möchte. Es ist einfach, `setInterval()` zu verwenden, aber es ist wichtig, das Intervall abzubrechen, wenn du es nicht mehr benötigst, um Speicherplatz zu sparen. React bietet [Lebenszyklusmethoden](/docs/react-component.html#the-component-lifecycle) an, die dich wissen lassen, wann eine Komponente im Begriff ist erstellt oder zerstört zu werden. Wir können ein Mixin erstellen, das diese Methoden verwendet, um eine einfache `setInterval()`-Funktion bereitzustellen, die automatisch bereinigt wird, wenn deine Komponente zerstört wird.
```

var createReactClass = require('create-react-class');

var TickTock = createReactClass({
  mixins: [SetIntervalMixin], // Verwende das Mixin
  getInitialState: function() {
    return {seconds: 0};
  },
  componentDidMount: function() {
    this.setInterval(this.tick, 1000); // Rufe eine Methode für das Mixin auf
  },
  tick: function() {
    this.setState({seconds: this.state.seconds + 1});
  },
  render: function() {
    return (
      <p>
        React wurde {this.state.seconds} Sekunden lang ausgeführt.
      </p>
    );
  }
});

ReactDOM.render(
  <TickTock />,

>   document.getElementById('beispiel')

Reviewer 2: ```suggestion
  document.getElementById('example')
```

We do not translate code

);
```

> Wenn eine Komponente mehrere Mixins verwendet und mehrere Mixins definieren dieselbe Lifecycle-Methode (d.H. Mehrere Mixins möchten eine Bereinigung durchführen, wenn die Komponente zerstört wird), werden alle Lifecycle-Methoden aufgerufen. Methoden, die für Mixins definiert wurden, werden in der Reihenfolge ausgeführt in der die Mixins aufgelistet wurden, gefolgt von einem Methodenaufruf für die Komponente.

Reviewer: ```suggestion
Wenn eine Komponente mehrere Mixins verwendet und mehrere Mixins definieren dieselbe Lifecycle-Methode (d. h. Mehrere Mixins möchten eine Bereinigung durchführen, wenn die Komponente zerstört wird), werden alle Lifecycle-Methoden aufgerufen. Methoden, die für Mixins definiert wurden, werden in der Reihenfolge ausgeführt in der die Mixins aufgelistet wurden, gefolgt von einem Methodenaufruf für die Komponente.
```

[d. h.](https://de.wiktionary.org/wiki/d._h.)

> Wenn eine Komponente mehrere Mixins verwendet und mehrere Mixins definieren dieselbe Lifecycle-Methode (d.H. Mehrere Mixins möchten eine Bereinigung durchführen, wenn die Komponente zerstört wird), werden alle Lifecycle-Methoden aufgerufen. Methoden, die für Mixins definiert wurden, werden in der Reihenfolge ausgeführt in der die Mixins aufgelistet wurden, gefolgt von einem Methodenaufruf für die Komponente.

Reviewer 2: ```suggestion
Wenn eine Komponente mehrere Mixins verwendet und mehrere Mixins definieren dieselbe Lifecycle-Methode (d.h. mehrere Mixins möchten eine Bereinigung durchführen, wenn die Komponente zerstört wird), werden alle Lifecycle-Methoden aufgerufen. Methoden, die für Mixins definiert wurden, werden in der Reihenfolge ausgeführt in der die Mixins aufgelistet wurden, gefolgt von einem Methodenaufruf für die Komponente.
```
